File: orion-code-svc/src/services/BuildRepository.ts
import { DatabasePool } from '../utils/database';
/**
 * BuildRepository - Database layer for Build operations
 * 
 * Handles PostgreSQL operations for build environments and build records
 */


export interface BuildEnvironment {
  id: string;
  tenant_id: string;
  name: string;
  type: string;
  image: string;
  description: string | null;
  config: Record<string, any>;
  status: string;
  created_at: Date;
  updated_at: Date;
}

export interface Build {
  id: string;
  tenant_id: string;
  project_id: string | null;
  pipeline_run_id: string | null;
  image: string | null;
  tag: string | null;
  status: string;
  source_ref: string | null;
  build_args: Record<string, any>;
  started_at: Date | null;
  completed_at: Date | null;
  duration_ms: number | null;
  error_message: string | null;
  created_at: Date;
}

export interface CreateBuildInput {
  tenant_id: string;
  project_id?: string;
  pipeline_run_id?: string;
  image?: string;
  tag?: string;
  source_ref?: string;
  build_args?: Record<string, any>;
}

export interface CreateBuildEnvironmentInput {
  tenant_id: string;
  name: string;
  type: string;
  image: string;
  description?: string;
  config?: Record<string, any>;
}

export interface UpdateBuildInput {
  status?: string;
  image?: string;
  tag?: string;
  error_message?: string;
}

interface FindAllOptions {
  tenantId?: string;
  projectId?: string;
  status?: string;
  limit?: number;
  offset?: number;
}

export class BuildRepository {
  constructor(private pool: DatabasePool) {}


  // ==================== Build Environments ====================

  /**
   * Find environment by ID
   */
  async findEnvironmentById(id: string): Promise<BuildEnvironment | null> {
    const result = await this.pool.query(
      'SELECT * FROM build_environments WHERE id = $1',
      [id]
    );
    return result.rows[0] || null;
  }

  /**
   * Find all environments
   */
  async findAllEnvironments(tenantId?: string): Promise<BuildEnvironment[]> {
    let query = 'SELECT * FROM build_environments';
    const params: any[] = [];

    if (tenantId) {
      params.push(tenantId);
      query += ' WHERE tenant_id = $1';
    }

    query += ' ORDER BY created_at DESC';

    const result = await this.pool.query(query, params);
    return result.rows;
  }

  /**
   * Create build environment
   */
  async createEnvironment(input: CreateBuildEnvironmentInput): Promise<BuildEnvironment> {
    const { tenant_id, name, type, image, description, config } = input;
    
    const result = await this.pool.query(
      `INSERT INTO build_environments (tenant_id, name, type, image, description, config)
       VALUES ($1, $2, $3, $4, $5, $6)
       RETURNING *`,
      [tenant_id, name, type, image, description || null, config || {}]
    );
    
    return result.rows[0];
  }

  /**
   * Update build environment
   */
  async updateEnvironment(id: string, input: Partial<CreateBuildEnvironmentInput & { status?: string }>): Promise<BuildEnvironment | null> {
    const updates: string[] = [];
    const params: any[] = [];
    let paramIndex = 1;

    if (input.name !== undefined) {
      params.push(input.name);
      updates.push(`name = $${paramIndex++}`);
    }

    if (input.image !== undefined) {
      params.push(input.image);
      updates.push(`image = $${paramIndex++}`);
    }

    if (input.description !== undefined) {
      params.push(input.description);
      updates.push(`description = $${paramIndex++}`);
    }

    if (input.config !== undefined) {
      params.push(JSON.stringify(input.config));
      updates.push(`config = $${paramIndex++}`);
    }

    if (input.status !== undefined) {
      params.push(input.status);
      updates.push(`status = $${paramIndex++}`);
    }

    if (updates.length === 0) {
      return this.findEnvironmentById(id);
    }

    params.push(id);
    
    const result = await this.pool.query(
      `UPDATE build_environments SET ${updates.join(', ')}, updated_at = NOW()
       WHERE id = $${paramIndex}
       RETURNING *`,
      params
    );

    return result.rows[0] || null;
  }

  /**
   * Delete environment (soft delete)
   */
  async deleteEnvironment(id: string): Promise<boolean> {
    const result = await this.pool.query(
      "UPDATE build_environments SET status = 'deleted', updated_at = NOW() WHERE id = $1",
      [id]
    );
    return (result.rowCount ?? 0) > 0;
  }

  // ==================== Builds ====================

  /**
   * Find build by ID
   */
  async findById(id: string): Promise<Build | null> {
    const result = await this.pool.query(
      'SELECT * FROM builds WHERE id = $1',
      [id]
    );
    return result.rows[0] || null;
  }

  /**
   * Find all builds with filtering
   */
  async findAll(options?: FindAllOptions): Promise<Build[]> {
    let query = 'SELECT * FROM builds';
    const params: any[] = [];
    const conditions: string[] = [];

    if (options?.tenantId) {
      params.push(options.tenantId);
      conditions.push(`tenant_id = $${params.length}`);
    }

    if (options?.projectId) {
      params.push(options.projectId);
      conditions.push(`project_id = $${params.length}`);
    }

    if (options?.status) {
      params.push(options.status);
      conditions.push(`status = $${params.length}`);
    }

    if (conditions.length > 0) {
      query += ' WHERE ' + conditions.join(' AND ');
    }

    query += ' ORDER BY created_at DESC';

    if (options?.limit) {
      params.push(options.limit);
      query += ` LIMIT $${params.length}`;
    }

    if (options?.offset) {
      params.push(options.offset);
      query += ` OFFSET $${params.length}`;
    }

    const result = await this.pool.query(query, params);
    return result.rows;
  }

  /**
   * Count builds
   */
  async count(options?: { tenantId?: string; status?: string }): Promise<number> {
    let query = 'SELECT COUNT(*) as count FROM builds';
    const params: any[] = [];

    if (options?.tenantId) {
      params.push(options.tenantId);
      query += ' WHERE tenant_id = $1';
      
      if (options?.status) {
        params.push(options.status);
        query += ' AND status = $2';
      }
    } else if (options?.status) {
      params.push(options.status);
      query += ' WHERE status = $1';
    }

    const result = await this.pool.query(query, params);
    return parseInt(result.rows[0].count, 10);
  }

  /**
   * Create a new build
   */
  async create(input: CreateBuildInput): Promise<Build> {
    const { tenant_id, project_id, pipeline_run_id, image, tag, source_ref, build_args } = input;
    
    const result = await this.pool.query(
      `INSERT INTO builds (tenant_id, project_id, pipeline_run_id, image, tag, source_ref, build_args, status)
       VALUES ($1, $2, $3, $4, $5, $6, $7, 'pending')
       RETURNING *`,
      [tenant_id, project_id || null, pipeline_run_id || null, image || null, tag || null, source_ref || null, build_args || {}]
    );
    
    return result.rows[0];
  }

  /**
   * Update build
   */
  async update(id: string, input: UpdateBuildInput): Promise<Build | null> {
    const updates: string[] = [];
    const params: any[] = [];
    let paramIndex = 1;

    if (input.status !== undefined) {
      params.push(input.status);
      updates.push(`status = $${paramIndex++}`);
    }

    if (input.image !== undefined) {
      params.push(input.image);
      updates.push(`image = $${paramIndex++}`);
    }

    if (input.tag !== undefined) {
      params.push(input.tag);
      updates.push(`tag = $${paramIndex++}`);
    }

    if (input.error_message !== undefined) {
      params.push(input.error_message);
      updates.push(`error_message = $${paramIndex++}`);
    }

    if (updates.length === 0) {
      return this.findById(id);
    }

    params.push(id);
    
    const result = await this.pool.query(
      `UPDATE builds SET ${updates.join(', ')}
       WHERE id = $${paramIndex}
       RETURNING *`,
      params
    );

    return result.rows[0] || null;
  }

  /**
   * Start build (set status to running)
   */
  async startBuild(id: string): Promise<Build | null> {
    const result = await this.pool.query(
      `UPDATE builds SET status = 'running', started_at = NOW()
       WHERE id = $1 RETURNING *`,
      [id]
    );
    return result.rows[0] || null;
  }

  /**
   * Complete build (success or failure)
   */
  async completeBuild(id: string, status: string, errorMessage?: string): Promise<Build | null> {
    const result = await this.pool.query(
      `UPDATE builds SET 
         status = $1, 
         completed_at = NOW(), 
         duration_ms = EXTRACT(EPOCH FROM (NOW() - started_at))::BIGINT * 1000,
         error_message = $2
       WHERE id = $3 
       RETURNING *`,
      [status, errorMessage || null, id]
    );
    return result.rows[0] || null;
  }

  /**
   * Get build by pipeline run
   */
  async findByPipelineRun(pipelineRunId: string): Promise<Build | null> {
    const result = await this.pool.query(
      'SELECT * FROM builds WHERE pipeline_run_id = $1 ORDER BY created_at DESC LIMIT 1',
      [pipelineRunId]
    );
    return result.rows[0] || null;
  }

  // ==================== Stats ====================

  /**
   * Get build statistics
   */
  async getBuildStats(tenantId?: string): Promise<{
    total: number;
    success: number;
    failed: number;
    running: number;
    pending: number;
    avgDuration: number;
  }> {
    let query = `SELECT 
      COUNT(*) as total,
      SUM(CASE WHEN status = 'success' THEN 1 ELSE 0 END) as success,
      SUM(CASE WHEN status = 'failed' THEN 1 ELSE 0 END) as failed,
      SUM(CASE WHEN status = 'running' THEN 1 ELSE 0 END) as running,
      SUM(CASE WHEN status = 'pending' THEN 1 ELSE 0 END) as pending,
      AVG(duration_ms) as avg_duration
     FROM builds`;
    const params: any[] = [];

    if (tenantId) {
      params.push(tenantId);
      query += ' WHERE tenant_id = $1';
    }

    const result = await this.pool.query(query, params);
    const row = result.rows[0];

    return {
      total: parseInt(row.total || '0', 10),
      success: parseInt(row.success || '0', 10),
      failed: parseInt(row.failed || '0', 10),
      running: parseInt(row.running || '0', 10),
      pending: parseInt(row.pending || '0', 10),
      avgDuration: parseFloat(row.avg_duration || '0'),
    };
  }
}

File: orion-build-svc-go/internal/repository/build_repository.go
package repository

import (
	"context"
	"database/sql"
	"fmt"
	"orion/build-svc-go/internal/models"
	"strings"
	"orion/go-common/pkg/database"
)

// BuildRepository handles all database operations for builds, environments, and artifacts.
type BuildRepository struct {
	db *database.DB
}

func NewBuildRepository(db *database.DB) *BuildRepository {
	return &BuildRepository{db: db}
}

// ==================== Builds ====================

func (r *BuildRepository) Create(ctx context.Context, b *models.Build) error {
	query := `
		INSERT INTO builds (tenant_id, project_id, pipeline_run_id, repo_id, branch, commit_sha,
			image, tag, source_ref, build_args, status)
		VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
		RETURNING id, created_at`
	return r.db.QueryRowContext(ctx, query,
		b.TenantID, b.ProjectID, b.PipelineRunID, b.RepoID, b.Branch, b.CommitSHA,
		b.Image, b.Tag, b.SourceRef, b.BuildArgs, b.Status,
	).Scan(&b.ID, &b.CreatedAt)
}

func (r *BuildRepository) GetByID(ctx context.Context, tenantID, id string) (*models.Build, error) {
	var b models.Build
	query := `SELECT id, tenant_id, project_id, pipeline_run_id, repo_id, branch, commit_sha,
		image, tag, source_ref, build_args, status, started_at, completed_at,
		duration_ms, error_message, logs, created_at
		FROM builds WHERE id = $1 AND tenant_id = $2`
	err := r.db.GetContext(ctx, &b, query, id, tenantID)
	if err != nil {
		return nil, fmt.Errorf("build not found: %w", err)
	}
	return &b, nil
}

func (r *BuildRepository) List(ctx context.Context, tenantID string, filter models.ListBuildsFilter, offset, limit int) ([]models.Build, error) {
	var builds []models.Build
	var conditions []string
	var args []interface{}
	argIdx := 1

	conditions = append(conditions, fmt.Sprintf("tenant_id = $%d", argIdx))
	args = append(args, tenantID)
	argIdx++

	if filter.ProjectID != "" {
		conditions = append(conditions, fmt.Sprintf("project_id = $%d", argIdx))
		args = append(args, filter.ProjectID)
		argIdx++
	}
	if filter.Status != "" {
		conditions = append(conditions, fmt.Sprintf("status = $%d", argIdx))
		args = append(args, filter.Status)
		argIdx++
	}

	where := "WHERE " + strings.Join(conditions, " AND ")

	query := fmt.Sprintf(`SELECT id, tenant_id, project_id, pipeline_run_id, repo_id, branch, commit_sha,
		image, tag, source_ref, build_args, status, started_at, completed_at,
		duration_ms, error_message, logs, created_at
		FROM builds %s ORDER BY created_at DESC LIMIT $%d OFFSET $%d`, where, argIdx, argIdx+1)
	args = append(args, limit, offset)

	err := r.db.SelectContext(ctx, &builds, query, args...)
	if err != nil {
		return nil, err
	}
	return builds, nil
}

func (r *BuildRepository) Update(ctx context.Context, b *models.Build) error {
	query := `
		UPDATE builds SET repo_id = $1, branch = $2, commit_sha = $3, status = $4,
			started_at = $5, completed_at = $6, image = $7, tag = $8,
			duration_ms = $9, error_message = $10, logs = $11
		WHERE id = $12 AND tenant_id = $13`
	_, err := r.db.ExecContext(ctx, query,
		b.RepoID, b.Branch, b.CommitSHA, b.Status,
		b.StartedAt, b.CompletedAt, b.Image, b.Tag,
		b.DurationMs, b.ErrorMessage, b.Logs,
		b.ID, b.TenantID,
	)
	return err
}

func (r *BuildRepository) UpdateStatus(ctx context.Context, tenantID, id, status string) error {
	query := `UPDATE builds SET status = $1 WHERE id = $2 AND tenant_id = $3`
	_, err := r.db.ExecContext(ctx, query, status, id, tenantID)
	return err
}

func (r *BuildRepository) Delete(ctx context.Context, tenantID, id string) error {
	query := `DELETE FROM builds WHERE id = $1 AND tenant_id = $2`
	_, err := r.db.ExecContext(ctx, query, id, tenantID)
	return err
}

func (r *BuildRepository) Count(ctx context.Context, tenantID string) (int, error) {
	var count int
	err := r.db.GetContext(ctx, &count, `SELECT COUNT(*) FROM builds WHERE tenant_id=$1`, tenantID)
	return count, err
}

// CountFiltered returns the count of builds matching the given filters, scoped to tenant.
func (r *BuildRepository) CountFiltered(ctx context.Context, tenantID string, filter models.ListBuildsFilter) (int, error) {
	var conditions []string
	var args []interface{}
	argIdx := 1

	conditions = append(conditions, fmt.Sprintf("tenant_id = $%d", argIdx))
	args = append(args, tenantID)
	argIdx++

	if filter.ProjectID != "" {
		conditions = append(conditions, fmt.Sprintf("project_id = $%d", argIdx))
		args = append(args, filter.ProjectID)
		argIdx++
	}
	if filter.Status != "" {
		conditions = append(conditions, fmt.Sprintf("status = $%d", argIdx))
		args = append(args, filter.Status)
		argIdx++
	}

	where := "WHERE " + strings.Join(conditions, " AND ")

	query := fmt.Sprintf("SELECT COUNT(*) FROM builds %s", where)
	var count int
	err := r.db.GetContext(ctx, &count, query, args...)
	return count, err
}

// StartBuild sets a build's status to 'running' and records started_at.
func (r *BuildRepository) StartBuild(ctx context.Context, tenantID, id string) (*models.Build, error) {
	query := `UPDATE builds SET status = 'running', started_at = NOW()
		WHERE id = $1 AND tenant_id = $2 AND status = 'pending'
		RETURNING id, tenant_id, project_id, pipeline_run_id, repo_id, branch, commit_sha,
			image, tag, source_ref, build_args, status, started_at, completed_at,
			duration_ms, error_message, logs, created_at`
	var b models.Build
	err := r.db.GetContext(ctx, &b, query, id, tenantID)
	if err != nil {
		return nil, fmt.Errorf("cannot start build: %w", err)
	}
	return &b, nil
}

// CompleteBuild marks a build as completed (success/failed/cancelled) with duration calculation.
func (r *BuildRepository) CompleteBuild(ctx context.Context, tenantID, id, status, errorMessage string) (*models.Build, error) {
	var errMsg interface{}
	if errorMessage != "" {
		errMsg = errorMessage
	}

	query := `UPDATE builds SET
		status = $1,
		completed_at = NOW(),
		duration_ms = EXTRACT(EPOCH FROM (NOW() - COALESCE(started_at, created_at)))::BIGINT * 1000,
		error_message = $2
		WHERE id = $3 AND tenant_id = $4
		RETURNING id, tenant_id, project_id, pipeline_run_id, repo_id, branch, commit_sha,
			image, tag, source_ref, build_args, status, started_at, completed_at,
			duration_ms, error_message, logs, created_at`
	var b models.Build
	err := r.db.GetContext(ctx, &b, query, status, errMsg, id, tenantID)
	if err != nil {
		return nil, fmt.Errorf("cannot complete build: %w", err)
	}
	return &b, nil
}

// FindByPipelineRun finds the most recent build for a pipeline run.
func (r *BuildRepository) FindByPipelineRun(ctx context.Context, tenantID, pipelineRunID string) (*models.Build, error) {
	var b models.Build
	query := `SELECT id, tenant_id, project_id, pipeline_run_id, repo_id, branch, commit_sha,
		image, tag, source_ref, build_args, status, started_at, completed_at,
		duration_ms, error_message, logs, created_at
		FROM builds WHERE pipeline_run_id = $1 AND tenant_id = $2
		ORDER BY created_at DESC LIMIT 1`
	err := r.db.GetContext(ctx, &b, query, pipelineRunID, tenantID)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, nil
		}
		return nil, err
	}
	return &b, nil
}

// GetBuildStats returns aggregated build statistics for a tenant.
func (r *BuildRepository) GetBuildStats(ctx context.Context, tenantID string) (*models.BuildStats, error) {
	var stats models.BuildStats
	query := `SELECT
		COUNT(*) AS total,
		COALESCE(SUM(CASE WHEN status = 'success' THEN 1 ELSE 0 END), 0) AS success,
		COALESCE(SUM(CASE WHEN status = 'failed' THEN 1 ELSE 0 END), 0) AS failed,
		COALESCE(SUM(CASE WHEN status = 'running' THEN 1 ELSE 0 END), 0) AS running,
		COALESCE(SUM(CASE WHEN status = 'pending' THEN 1 ELSE 0 END), 0) AS pending,
		COALESCE(AVG(duration_ms), 0) AS avg_duration
		FROM builds WHERE tenant_id = $1`
	err := r.db.GetContext(ctx, &stats, query, tenantID)
	if err != nil {
		return nil, err
	}
	return &stats, nil
}

// ==================== Build Environments ====================

func (r *BuildRepository) CreateEnvironment(ctx context.Context, env *models.BuildEnvironment) error {
	query := `INSERT INTO build_environments (tenant_id, name, type, image, description, config)
		VALUES ($1, $2, $3, $4, $5, $6)
		RETURNING id, created_at, updated_at`
	return r.db.QueryRowContext(ctx, query,
		env.TenantID, env.Name, env.Type, env.Image, env.Description, env.Config,
	).Scan(&env.ID, &env.CreatedAt, &env.UpdatedAt)
}

func (r *BuildRepository) GetEnvironmentByID(ctx context.Context, tenantID, id string) (*models.BuildEnvironment, error) {
	var env models.BuildEnvironment
	query := `SELECT id, tenant_id, name, type, image, description, config, status, created_at, updated_at
		FROM build_environments WHERE id = $1 AND tenant_id = $2`
	err := r.db.GetContext(ctx, &env, query, id, tenantID)
	if err != nil {
		return nil, fmt.Errorf("environment not found: %w", err)
	}
	return &env, nil
}

func (r *BuildRepository) ListEnvironments(ctx context.Context, tenantID string) ([]models.BuildEnvironment, error) {
	var envs []models.BuildEnvironment
	query := `SELECT id, tenant_id, name, type, image, description, config, status, created_at, updated_at
		FROM build_environments WHERE tenant_id = $1 AND status != 'deleted' ORDER BY created_at DESC`
	err := r.db.SelectContext(ctx, &envs, query, tenantID)
	if err != nil {
		return nil, err
	}
	return envs, nil
}

func (r *BuildRepository) UpdateEnvironment(ctx context.Context, env *models.BuildEnvironment) error {
	query := `UPDATE build_environments SET name = $1, type = $2, image = $3,
		description = $4, config = $5, status = $6, updated_at = NOW()
		WHERE id = $7 AND tenant_id = $8`
	_, err := r.db.ExecContext(ctx, query,
		env.Name, env.Type, env.Image, env.Description, env.Config, env.Status, env.ID, env.TenantID,
	)
	return err
}

func (r *BuildRepository) DeleteEnvironment(ctx context.Context, tenantID, id string) error {
	query := `UPDATE build_environments SET status = 'deleted', updated_at = NOW() WHERE id = $1 AND tenant_id = $2`
	_, err := r.db.ExecContext(ctx, query, id, tenantID)
	return err
}

// ==================== Artifacts ====================

func (r *BuildRepository) CreateArtifact(ctx context.Context, a *models.Artifact) error {
	query := `INSERT INTO artifacts (tenant_id, name, type, storage_type, storage_path,
		size_bytes, checksum_sha256, run_id, stage_id, expires_at, metadata)
		VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
		RETURNING id, created_at, updated_at`
	return r.db.QueryRowContext(ctx, query,
		a.TenantID, a.Name, a.Type, a.StorageType, a.StoragePath,
		a.SizeBytes, a.ChecksumSHA256, a.RunID, a.StageID, a.ExpiresAt, a.Metadata,
	).Scan(&a.ID, &a.CreatedAt, &a.UpdatedAt)
}

func (r *BuildRepository) GetArtifactByID(ctx context.Context, tenantID, id string) (*models.Artifact, error) {
	var a models.Artifact
	query := `SELECT id, tenant_id, name, type, storage_type, storage_path, size_bytes,
		checksum_sha256, run_id, stage_id, expires_at, downloaded_count, metadata, created_at, updated_at
		FROM artifacts WHERE id = $1 AND tenant_id = $2`
	err := r.db.GetContext(ctx, &a, query, id, tenantID)
	if err != nil {
		return nil, fmt.Errorf("artifact not found: %w", err)
	}
	return &a, nil
}

func (r *BuildRepository) ListArtifacts(ctx context.Context, tenantID string, filter models.ListArtifactFilter, offset, limit int) ([]models.Artifact, error) {
	var artifacts []models.Artifact
	var conditions []string
	var args []interface{}
	argIdx := 1

	conditions = append(conditions, fmt.Sprintf("tenant_id = $%d", argIdx))
	args = append(args, tenantID)
	argIdx++

	if filter.RunID != "" {
		conditions = append(conditions, fmt.Sprintf("run_id = $%d", argIdx))
		args = append(args, filter.RunID)
		argIdx++
	}
	if filter.StageID != "" {
		conditions = append(conditions, fmt.Sprintf("stage_id = $%d", argIdx))
		args = append(args, filter.StageID)
		argIdx++
	}
	if filter.Type != "" {
		conditions = append(conditions, fmt.Sprintf("type = $%d", argIdx))
		args = append(args, filter.Type)
		argIdx++
	}

	where := "WHERE " + strings.Join(conditions, " AND ")
	query := fmt.Sprintf(`SELECT id, tenant_id, name, type, storage_type, storage_path, size_bytes,
		checksum_sha256, run_id, stage_id, expires_at, downloaded_count, metadata, created_at, updated_at
		FROM artifacts %s ORDER BY created_at DESC LIMIT $%d OFFSET $%d`, where, argIdx, argIdx+1)
	args = append(args, limit, offset)

	err := r.db.SelectContext(ctx, &artifacts, query, args...)
	if err != nil {
		return nil, err
	}
	return artifacts, nil
}

func (r *BuildRepository) DeleteArtifact(ctx context.Context, tenantID, id string) error {
	_, err := r.db.ExecContext(ctx, `DELETE FROM artifacts WHERE id = $1 AND tenant_id = $2`, id, tenantID)
	return err
}

func (r *BuildRepository) IncrementDownloadCount(ctx context.Context, tenantID, id string) error {
	_, err := r.db.ExecContext(ctx,
		`UPDATE artifacts SET downloaded_count = downloaded_count + 1, updated_at = NOW() WHERE id = $1 AND tenant_id = $2`, id, tenantID)
	return err
}

func (r *BuildRepository) CleanupExpiredArtifacts(ctx context.Context, tenantID string) (int, error) {
	result, err := r.db.ExecContext(ctx,
		`DELETE FROM artifacts WHERE tenant_id = $1 AND expires_at IS NOT NULL AND expires_at < NOW()`, tenantID)
	if err != nil {
		return 0, err
	}
	rows, _ := result.RowsAffected()
	return int(rows), nil
}

func (r *BuildRepository) CleanupArtifactsByRun(ctx context.Context, tenantID, runID string) (int, error) {
	result, err := r.db.ExecContext(ctx, `DELETE FROM artifacts WHERE tenant_id = $1 AND run_id = $2`, tenantID, runID)
	if err != nil {
		return 0, err
	}
	rows, _ := result.RowsAffected()
	return int(rows), nil
}
